cpp/src/parquet/encoding.cc

```diff
@@ -2501,20 +2596,13 @@ void ByteStreamSplitDecoder<DType>::SetData(int num_values, const uint8_t* data,
 
 template <typename DType>
 int ByteStreamSplitDecoder<DType>::Decode(T* buffer, int max_values) {
-  constexpr size_t num_streams = sizeof(T);
-  const int values_to_decode = std::min(num_values_, max_values);
-  const int num_decoded_previously = num_values_in_buffer - num_values_;
-  for (int i = 0; i < values_to_decode; ++i) {
-    uint8_t gathered_byte_data[num_streams];
-    for (size_t b = 0; b < num_streams; ++b) {
-      const size_t byte_index = b * num_values_in_buffer + num_decoded_previously + i;
-      gathered_byte_data[b] = data_[byte_index];
-    }
-    buffer[i] = arrow::util::SafeLoadAs<T>(&gathered_byte_data[0]);
-  }
-  num_values_ -= values_to_decode;
-  len_ -= sizeof(T) * values_to_decode;
-  return values_to_decode;
+#if defined(ARROW_HAVE_SSE4_2) || defined(ARROW_HAVE_SSE2)
+  constexpr bool has_sse2_or_higher = true;
+#else
+  constexpr bool has_sse2_or_higher = false;
+#endif
```

Comment:
As discussed on the mailing list we probably are going to want `encoding_sse42.cc` that is always compiled with `-msse4.2`, and then use runtime dispatching to select the best implementation for the current CPU. Otherwise these optimizations will never make it to many users since packages have to be made compatible on a wide array of CPUs

Reply:
We always compile with SSE4.2 enabled AFAIR (on x86 of course).
I think runtime dispatching is a red herring here: we'll have it when someone implements it.

Reply:
Ack